gloo/mpi/context.cc

```diff
@@ -33,6 +34,53 @@ static int MPICommRank(const MPI_Comm& comm) {
   return comm_rank;
 }
 
+MPIScope::MPIScope() {
+  auto rv = MPI_Init(nullptr, nullptr);
+  GLOO_ENFORCE_EQ(rv, MPI_SUCCESS);
+}
+
+MPIScope::~MPIScope() {
+  auto rv = MPI_Finalize();
+  GLOO_ENFORCE_EQ(rv, MPI_SUCCESS);
+}
+
+namespace {
+
+std::shared_ptr<MPIScope> getMPIScope() {
+  static std::once_flag once;
+
+  // Use weak pointer so that the initializer is destructed when the
+  // last context referring to it is destructed, not when statics
+  // are destructed on program termination.
+  static std::weak_ptr<MPIScope> wptr;
+  std::shared_ptr<MPIScope> sptr;
+
+  // Create MPIScope only once
+  std::call_once(once, [&]() {
+      sptr = std::make_shared<MPIScope>();
+      wptr = sptr;
+    });
+
+  // Grab reference to MPIScope.
+  // This is a potential race since contexts may be destructed while
+  // new ones are constructed. No easy way to fix this.
+  if (!sptr) {
+    GLOO_ENFORCE(!wptr.expired(), "Cannot initialize MPIScope twice");
+    sptr = wptr.lock();
```

Comment:
Looks like you can address this race condition by checking whether the result of `lock()` is empty:
```cpp
    if (auto observe = weak.lock()) {
        std::cout << "\tobserve() able to lock weak_ptr<>, value=" << *observe << "\n";
    } else {
        std::cout << "\tobserve() unable to lock weak_ptr<>\n";
    }
```
http://en.cppreference.com/w/cpp/memory/weak_ptr/lock

Reply:
Thanks, updated to only use `.lock()`.